llm/orchestrator-server/src/main/python/app/src/llm_orchestrator/services/langchain/rag_chain.py

```diff
@@ -12,39 +12,92 @@
+    get_vector_store_factory,
 )
 
 
+@opensearch_exception_handler
+@openai_exception_handler(provider='OpenAI or AzureOpenAIService')
 def execute_qa_chain(query: RagQuery, debug: bool) -> RagResponse:
-    llm_factory = get_llm_factory(query.question_answering_llm_setting)
-    em_factory = get_em_factory(query.embedding_question_em_setting)
+    llm_factory = get_llm_factory(setting=query.question_answering_llm_setting)
+    em_factory = get_em_factory(setting=query.embedding_question_em_setting)
+    vector_store_factory = get_vector_store_factory(
+        vector_store_provider=VectorStoreProvider.OPEN_SEARCH,
+        embedding_function=em_factory.get_embedding_model(),
+        index_name=query.index_name,
+    )
 
-    # Instantiate LangChain, using :
-    # llm_factory.get_language_model()
-    # em_factory.get_embedding_model()
+    chat = ConversationalRetrievalChain.from_llm(
+        llm=llm_factory.get_language_model(),
+        retriever=vector_store_factory.get_vector_store().as_retriever(
+            search_kwargs=query.document_search_params.to_dict()
+        ),
+        return_source_documents=True,
+        return_generated_question=True,
+        combine_docs_chain_kwargs={
+            'prompt': PromptTemplate(
+                template=llm_factory.setting.prompt,
+                input_variables=['context', 'question'],
+            )
+        },
+    )
+
+    message_history = ChatMessageHistory()
+    for msg in query.history:
+        if ChatMessageType.HUMAN == msg.type:
+            message_history.add_user_message(msg.text)
+        else:
+            message_history.add_ai_message(msg.text)
+
+    records_callback_handler = RetrieverJsonCallbackHandler()
+
+    response = chat(
+        inputs={
+            'question': query.question,
+            'chat_history': message_history.messages,
+        },
+        callbacks=([], [records_callback_handler])[debug],
+    )
 
-    # Fake answer
     return RagResponse(
         answer=TextWithFootnotes(
-            text='{} - {} - {}'.format(
-                llm_factory.get_language_model(),
-                em_factory.get_embedding_model(),
-                debug,
+            text=response['answer'],
+            footnotes=set(
+                map(
+                    lambda doc: Footnote(
+                        identifier=f'{doc.metadata["id"]}',
+                        title=doc.metadata['title'],
+                        url=doc.metadata['url'],
+                    ),
+                    response['source_documents'],
+                )
             ),
-            footnotes=[
-                Footnote(
-                    identifier='1',
-                    title='Tock Documentation',
-                    url='https://doc.tock.ai/tock',
-                ),
-                Footnote(identifier='2', title='Other source'),
-            ],
         ),
-        debug=[],
+        debug=(None, records_callback_handler.show_records())[debug],
```

Comment:
This isn't easy to read and also not optimal as you are always build a tuple even, prefer a clear ternary operator : 
```suggestion
        debug= records_callback_handler.show_records() if debug else None,
```